# Genomic Data - DemoCodingVsIntergenomicSeqs

This is one of the genomic datasets taken from [here](https://github.com/ML-Bioinfo-CEITEC/genomic_benchmarks).
The classification task is evaluated using the _SeqRep_ package.

You can [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/MIR-MU/seqrep/blob/main/examples/DemoCodingVsIntergenomicSeqs.ipynb)
or
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/MIR-MU/seqrep/main?labpath=examples%2FDemoCodingVsIntergenomicSeqs.ipynb).

## Install _SeqRep_ Package

In [ ]:
!pip install seqrep

     |████████████████████████████████| 115 kB 4.3 MB/s 
     |████████████████████████████████| 303 kB 28.5 MB/s 
     |████████████████████████████████| 15.3 MB 13.1 MB/s 
  Created wheel for pandas-ta: filename=pandas_ta-0.3.14b0-py3-none-any.whl size=218923 sha256=d90adf009c3bd9c58e9bf7c917d123c434d46c60e281cde8b2af618e79b796c4
  Stored in directory: /root/.cache/pip/wheels/0b/81/f0/cca85757840e4616a2c6b9fe12569d97d324c27cac60724c58
  Created wheel for ta: filename=ta-0.9.0-py3-none-any.whl size=28908 sha256=2646be1de16d3448fe3b4b22b48a21913981ed952fdc8cf92b503bb8f8a62a1e
  Stored in directory: /root/.cache/pip/wheels/72/78/64/cc1c01506a1010a9845e9bd7c69333730f7174661228ea4f98
Successfully built pandas-ta ta
  Attempting uninstall: numpy
    Found existing installation: numpy 1.19.5
    Uninstalling numpy-1.19.5:
      Successfully uninstalled numpy-1.19.5
  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.0
    Uninstalling joblib-1.1.0:
      Successfully 

## Import Needed Packages

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.neural_network import MLPClassifier

!pip install icecream
from icecream import ic

from seqrep import *
from seqrep.feature_engineering import *
from seqrep.labeling import *
from seqrep.splitting import *
from seqrep.scaling import *
from seqrep.feature_reduction import *
from seqrep.evaluation import *
from seqrep.pipeline_evaluation import *

## Load or Download Data

In [ ]:
!pip install genomic-benchmarks

     |████████████████████████████████| 2.3 MB 3.8 MB/s 
     |████████████████████████████████| 596 kB 47.9 MB/s 
     |████████████████████████████████| 271 kB 51.2 MB/s 
     |████████████████████████████████| 94 kB 3.2 MB/s 
  Created wheel for genomic-benchmarks: filename=genomic_benchmarks-0.0.6-py3-none-any.whl size=17526 sha256=6be908cdf272ca41ae55ef6246d8add0e71846597c6d5a7affaa84e28d53a289
  Stored in directory: /root/.cache/pip/wheels/a7/0f/e4/f723fa5070923dc495b63e02600faf64bf76dbdf422ebdeb66
Successfully built genomic-benchmarks
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [ ]:
from genomic_benchmarks.data_check import list_datasets

list_datasets()

['human_enhancers_cohn',
 'human_enhancers_ensembl',
 'human_nontata_promoters',
 'demo_human_or_worm',
 'demo_mouse_enhancers',
 'demo_coding_vs_intergenomic_seqs']

In [ ]:
["".join([w.capitalize() for w in dat.split("_")]) for dat in list_datasets()]

['HumanEnhancersCohn',
 'HumanEnhancersEnsembl',
 'HumanNontataPromoters',
 'DemoHumanOrWorm',
 'DemoMouseEnhancers',
 'DemoCodingVsIntergenomicSeqs']

In [ ]:
# demo_coding_vs_intergenomic_seqs
from genomic_benchmarks.dataset_getters.pytorch_datasets import (
    DemoCodingVsIntergenomicSeqs,
)

X_train = DemoCodingVsIntergenomicSeqs(split="train", version=0)
X_test = DemoCodingVsIntergenomicSeqs(split="test", version=0)


y_train = pd.Series([y for _, y in X_train])
X_train = pd.DataFrame([x for x, _ in X_train], columns=["genom"])
y_test = pd.Series([y for _, y in X_test])
X_test = pd.DataFrame([x for x, _ in X_test], columns=["genom"])

Unzipping...Done.


In [ ]:
## Random shuffling
# idx = np.random.permutation(len(y_train))
# X_train = X_train.iloc[idx, :].reset_index(drop=True)
# y_train = y_train.iloc[idx].reset_index(drop=True)

# idx = np.random.permutation(len(y_test))
# X_test = X_test.iloc[idx, :].reset_index(drop=True)
# y_test = y_test.iloc[idx].reset_index(drop=True)

X_train.join(pd.DataFrame(y_train, columns=["label"]))

genom  label
0      GCCACCTTCCCCGCTTCAGGCCTACAATGGCCATCATCAGCTTTCT...      0
1      AGAAACGTCTTTGTGATGTTTGCATTCAACTCATAGAGTTGAACAT...      0
2      CATTCTCAGAAACGTCTTTGTGATGTTTGCATTCAACTCATAGAGT...      0
3      GATTTTAAAAGATTGATTTTTTTTTAAAGAAATGAATCCCAAATCT...      0
4      GGATGTTTCAATTGAAGTCCCAGTGTTGAACATTCCCTTTCATAGA...      0
...                                                  ...    ...
74995  GGCTGGAAGAGGATCTTGGGCGCCGCCAGGTTCTGTGGACAATCAC...      1
74996  CCGCCGCCTCTCCTTGCTGCAGCCATGGAGTCTTCCACTTTCGCCT...      1
74997  AAGTGAAAAGGAGACACGGAATAAACTCATTGAGTCAGATCTGAAC...      1
74998  TGGGTGGCCAGTACTTGCGGGTGGGCAAGGCTGTCACACCGCCCAT...      1
74999  TCCATTCAGAACGTGTCTGTGGTCTTCAAGGGGACCCTGAAGTATG...      1

[75000 rows x 2 columns]

## Run Pipeline Evaluation

In [ ]:
# This DataFrame collects the results of various runs for comparison.

# Uncomment following line if you want to clear the DataFrame with the results.
# del results_for_comparison

try:
    results_for_comparison
except NameError:
    print("Create new empty DataFrame.")
    results_for_comparison = pd.DataFrame()
else:
    print("DataFrame already exist!")

Create new empty DataFrame.


In [ ]:
class SubstringsExtractor(FeatureExtractor):

    def __init__(self, substrings: List, columns_to_apply: Union[str, List[str]] = None,
                 return_original_columns: bool = False, normalize: bool = True,
                 verbose: bool = True, inplace: bool = False):
        self.substrings = substrings
        self.columns_to_apply = columns_to_apply
        self.return_original_columns = return_original_columns
        self.normalize = normalize
        self.verbose = verbose
        self.inplace = inplace

    def fit(self, X, y=None):
        if self.verbose:
            print(f"\tNumber of substrings BEFORE fit: {len(self.substrings)}")
        if not self.columns_to_apply:
            self.columns_to_apply = X.columns
        if isinstance(self.columns_to_apply, str):
            columns_to_apply = [self.columns_to_apply]
        
        tmp =
        for c in self.columns_to_apply:
            tmp += "@".join(X[c])
        self.substrings = [s for s in tqdm(self.substrings, leave=False, desc="Fitting SubstringsExtractor") if s in tmp]
        del tmp

        if self.verbose:
            print(f"\tNumber of substrings AFTER fit:  {len(self.substrings)}")
        return self

    def transform(self, X):
        if not self.inplace:
            X = X.copy()
        for column in tqdm(self.columns_to_apply, leave=False, desc="Transforming SubstringsExtractor - columns"):
            col_pref = column + "_" if len(self.columns_to_apply) > 1 else ""
            for substr in tqdm(self.substrings, leave=False, desc="Transforming SubstringsExtractor - substrings"):
                X.loc[:, f"{col_pref}count-{substr}"] = X[column].str.count(substr) / X[column].str.len() if self.normalize else 1

        if self.return_original_columns:
            return X
        return X.drop(columns=self.columns_to_apply)


In [ ]:
from itertools import product

perms = [
    "".join(p)
    for i in range(1, 7)
    for p in product("ACTG", repeat=i)
    if len(p) < 5 or len(set(p)) > 2
]
ic(len(perms))
# ic(perms[:10])

ic| len(perms): 4900


4900

In [ ]:
%%capture --no-stdout --no-display
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score

run_identification = f"{len(perms)} substrings"

# 1. step - define your pipeline
pipe = Pipeline(
    [
        (
            "fext_substr",
            SubstringsExtractor(
                substrings=perms,
            ),
        ),
        ("scale_u", UniversalScaler(scaler=MinMaxScaler())),
    ]
)

# 2. step - define your workflow
pipe_eval = PipelineEvaluator(
    pipeline=pipe,
    model=MLPClassifier(
        hidden_layer_sizes=(128, 32, 8),
        batch_size=32,
    ),
    evaluator=SequentialEvaluator(
        [
            ClassificationEvaluator(),
            UniversalEvaluator(metrics=[f1_score]),
        ]
    ),
)
# 3. step
pipe_eval.X_train = X_train.copy()
pipe_eval.y_train = y_train.copy()
pipe_eval.X_test = X_test.copy()
pipe_eval.y_test = y_test.copy()

result = pipe_eval.run()

results_for_comparison = results_for_comparison.append(
    pd.Series(result, name=run_identification),
)

09:36:56.395 Fitting pipeline
	Number of substrings BEFORE fit: 4900


Fitting SubstringsExtractor:   0%|          | 0/4900 [00:00<?, ?it/s]

	Number of substrings AFTER fit:  4900


Transforming SubstringsExtractor - columns:   0%|          | 0/1 [00:00<?, ?it/s]

Transforming SubstringsExtractor - substrings:   0%|          | 0/4900 [00:00<?, ?it/s]

09:52:33.019 Applying pipeline transformations


Transforming SubstringsExtractor - columns:   0%|          | 0/1 [00:00<?, ?it/s]

Transforming SubstringsExtractor - substrings:   0%|          | 0/4900 [00:00<?, ?it/s]

09:57:54.700 	Original shape:		(75000, 4900); 
		shape after removing NaNs: (75000, 4900).
09:57:55.856 	Original shape:		(25000, 4900); 
		shape after removing NaNs: (25000, 4900).
09:57:55.856 Fitting model
12:43:54.484 Predicting
12:44:09.092 Evaluating predictions
[[11103  1397]
 [ 1170 11330]] 
 89.732 % accuracy
 89.02333621434745 % precision of 1 classes
 90.64 % recall of 1 classes

              precision    recall  f1-score   support

           0       0.90      0.89      0.90     12500
           1       0.89      0.91      0.90     12500

    accuracy                           0.90     25000
   macro avg       0.90      0.90      0.90     25000
weighted avg       0.90      0.90      0.90     25000

f1_score:
	0.8982439449795854


In [ ]:
results_for_comparison

accuracy_score  ...  f1_score
97 substrings           0.85152  ...  0.849253
4900 substrings         0.89732  ...  0.898244

[2 rows x 5 columns]

| Dataset                          |   Accuracy |   F1 score |   |
|:---------------------------------|-----------:|-----------:|-----------:|
| demo_coding_vs_intergenomic_seqs |    85.152 |     84.9253 | 97 substrings - combinations_with_replacement	 |
| demo_coding_vs_intergenomic_seqs |    84.556 |     84.7866 |  |
| demo_coding_vs_intergenomic_seqs |    89.732 |     89.8244 | 4900 substrings |